File: 2023-1/Basic/bjgood01/Recursion/11729.py
n = int(input())  # 원판의 개수
result = 0        # 옮김 횟수
def hanoi(n, start, to, end):
    if n == 1:
        print(start, end)
    else:
        hanoi(n-1, start, end, to) # n-1개 원판을 start에서 end를 거쳐 to로 옮김
        print(start, end)          # 가장 아래 원판 보냄
        hanoi(n-1, to, start, end) # n-1개 원판을 to에서 start를 거쳐 end로 옮김

# 원판 1개면 옮긴 횟수 = 2^1 -1 => 1번
# 원판 2개면 옮긴 횟수 = 2^2 -1 => 3번
# 원판 3개면 옮긴 횟수 = 2^3 -1 => 7번
#...
result = 2 ** n -1
print(result)
hanoi(n, 1, 2, 3)

File: 2023-2/Basic/user20230912/Recursion/11729.cpp
#include <iostream>
#include <queue>

std::queue<std::pair< int, int >> q;

int count;

void hanoi(int n, int from, int temp, int to);

int main()
{
    int input;

    std::cin >> input;

    hanoi(input, 1, 2, 3);

    std::cout << count << '\n';
    while (!q.empty()) {
        std::cout << q.front().first << ' ' << q.front().second << '\n';
        q.pop();
    }

}

void hanoi(int n, int from, int temp, int to) {

    if (n == 1) {
        q.push({ from, to });
        count++;
    }
    else {
        hanoi(n - 1, from, to, temp);
        q.push({ from,to }); count++;
        hanoi(n - 1, temp, from, to);
    }

}
